### PR TITLE
Add support for `no_new_privs` via AllowPrivilegeEscalation

### LINKED ISSUE #38417: Add support for "no new privileges"
Support for setting [NO_NEW_PRIVS](https://www.kernel.org/doc/Documentation/prctl/no_new_privs.txt) was added in docker 1.11 (https://github.com/docker/docker/issues/20329). This feature will be much more useful once user namespace support is added (https://github.com/kubernetes/kubernetes/pull/34569), but I believe it is also relevant when paired with seccomp or LSM (SELinux or AppArmor).

/cc @mrunalp @kubernetes/sig-node

### PR DESCRIPTION
**Pull request by jessfraz**
**What this PR does / why we need it**:
Implements kubernetes/community#639
Fixes #38417

Adds `AllowPrivilegeEscalation` and `DefaultAllowPrivilegeEscalation` to `PodSecurityPolicy`.
Adds `AllowPrivilegeEscalation` to container `SecurityContext`.

Adds the proposed behavior to `kuberuntime`, `dockershim`, and `rkt`. Adds a bunch of unit tests to ensure the desired default behavior and that when `DefaultAllowPrivilegeEscalation` is explicitly set.

Tests pass locally with docker and rkt runtimes. There are also a few integration tests with a `setuid` binary for sanity.

**Release note**:

```release-note
Adds AllowPrivilegeEscalation to control whether a process can gain more privileges than it's parent process
```

### CODE REVIEW

**Comment by dims** [comment]
@jessfraz : sorry for a dumb question. Is this feature meant to only be active/available when the kubelet `--allow-privileged` is true?

**Comment by jessfraz** [comment]
@dims there is a chart in the proposal that might help, `no_new_privs` will be added when allowPrivilegeEscalation is false, by default this will be when a container is not privileged, adding CAP{_SYS_ADMIN, or when the uid!=0

**Comment by dims** [comment]
@jessfraz got it. thanks!

**Comment by jessfraz** [comment]
Tests pass locally with `docker` and `rkt` runtimes.

File: test/e2e_node/image_list.go [hunk 1]
```diff
@@ -53,6 +53,7 @@ var NodeImageWhiteList = sets.NewString(
 	"gcr.io/google_containers/nginx-slim:0.7",
 	"gcr.io/google_containers/serve_hostname:v1.4",
 	"gcr.io/google_containers/netexec:1.7",
+	"r.j3ss.co/no_new_privs:latest",
```
Comment (jessfraz) [inline test/e2e_node/image_list.go]: I will switch this to `google_containers` before merge don't worry

**Comment by pweil-** [comment]
@php-coder

File: pkg/securitycontext/util.go [hunk 2]
```diff
@@ -134,6 +134,11 @@ func DetermineEffectiveSecurityContext(pod *v1.Pod, container *v1.Container) *v1
 		*effectiveSc.ReadOnlyRootFilesystem = *containerSc.ReadOnlyRootFilesystem
 	}
 
+	if containerSc.AllowPrivilegeEscalation != nil {
```
Comment (pweil-) [inline pkg/securitycontext/util.go]: :+1: on finding this, the `DetermineEffectiveSecurityContext` will cause subtle issues in PSP if not handled (and I've forgotten it a number of times)

Comment (jessfraz) [inline pkg/securitycontext/util.go]: I felt relief after I found it too hahahahaaha

**Comment by jessfraz** [comment]
The test failure is not the test I added so whoop lol *cries*

File: test/e2e_node/image_list.go (hunk 1 above)
Comment (jessfraz) [inline test/e2e_node/image_list.go]: fixed the image to be on google container registry

**Comment by jessfraz** [comment]
@k8s-bot pull-kubernetes-node-e2e test this

**Comment by jessfraz** [comment]
Ok yay they all passed

**Comment by jessfraz** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

File: test/images/nonewprivs/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM alpine:latest
+
+COPY nnp /usr/local/bin/nnp
+RUN chmod +s /usr/local/bin/nnp
```
Comment (lucab) [inline test/images/nonewprivs/Dockerfile]: I see you want this compiled as a static ELF in the Makefile. If that's the case, can we avoid an alpine dependency and a spurious layer and just use `FROM: scratch` plus `sudo chmod` on the host side? (Other test containers in here seems to do so)

File: pkg/apis/extensions/v1beta1/types.go [hunk 4]
```diff
@@ -933,6 +933,15 @@ type PodSecurityPolicySpec struct {
 	// host paths may be used.
 	// +optional
 	AllowedHostPaths []string `json:"allowedHostPaths,omitempty" protobuf:"bytes,15,opt,name=allowedHostPaths"`
+	// DefaultAllowPrivilegeEscalation controls the default setting for whether a
+	// process can gain more privileges than it's parent process.
+	// // +optional
```
Comment (lucab) [inline pkg/apis/extensions/v1beta1/types.go]: I hope this is the source location, because those are propagated to all auto-generated snippets:
 * s/it's parent/its parent/
 * Spurious double slash before the optional annotation

The first typo is also in PR comment / release note, but I can't comment directly there :)

File: test/e2e_node/security_context_test.go [hunk 5]
```diff
@@ -274,4 +276,84 @@ var _ = framework.KubeDescribe("Security Context", func() {
 		})
 	})
 
+	Context("when creating containers with AllowPrivilegeEscalation", func() {
+		// parse the docker version
+		out, err := exec.Command("docker", "-v").CombinedOutput()
```
Comment (lucab) [inline test/e2e_node/security_context_test.go]: There is a `ContainerRuntime` in `framework/TestContext` to make this conditional and effective only for docker test environments.

**Review by lucab — CHANGES_REQUESTED** [review]

File: test/images/nonewprivs/Dockerfile (hunk 3 above)
Comment (jessfraz) [inline test/images/nonewprivs/Dockerfile]: Can update

File: pkg/apis/extensions/v1beta1/types.go (hunk 4 above)
Comment (jessfraz) [inline pkg/apis/extensions/v1beta1/types.go]: Gotcha thanks

File: test/e2e_node/security_context_test.go (hunk 5 above)
Comment (jessfraz) [inline test/e2e_node/security_context_test.go]: Thanks I looked all over

File: pkg/security/podsecuritypolicy/provider.go [hunk 6]
```diff
@@ -310,6 +316,10 @@ func (s *simpleProvider) ValidateContainerSecurityContext(pod *api.Pod, containe
 		}
 	}
 
+	if !s.psp.Spec.AllowPrivilegeEscalation && sc.AllowPrivilegeEscalation != nil && *sc.AllowPrivilegeEscalation {
```
Comment (timstclair) [inline pkg/security/podsecuritypolicy/provider.go]: I think this always happens after defaulting, which means we should require that `AllowPrivilegeEscalation != nil`. Otherwise if the admin forgets to set a default and the user leaves this empty, that's equivalent to allowing privilige escalation in certain scenarios (IIUC). Alternatively, we could default to false if it's prohibited by the psp and no default is set.

File: pkg/securitycontext/util.go [hunk 7]
```diff
@@ -232,3 +237,51 @@ func internalSecurityContextFromPodSecurityContext(pod *api.Pod) *api.SecurityCo
 
 	return synthesized
 }
+
+// AddNoNewPrivileges returns if we should add the no_new_privs option. This will return true if:
+// 1) the container is not privileged
+// 2) CAP_SYS_ADMIN is not being added
+// 3) if podSecurityPolicy.DefaultAllowPrivilegeEscalation is:
+//		- nil, then return true if uid != 0
```
Comment (timstclair) [inline pkg/securitycontext/util.go]: This will still break SUID binaries if the uid is set in the Dockerfile rather than through k8s, right?

File: test/images/nonewprivs/Makefile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+.PHONY: all image push clean
+
+TAG = 1.0
+PREFIX = gcr.io/google_containers
```
Comment (timstclair) [inline test/images/nonewprivs/Makefile]: nit: s/google_containers/google-containers/

File: test/e2e_node/security_context_test.go (hunk 5 above)
Comment (timstclair) [inline test/e2e_node/security_context_test.go]: Better yet, copy [SkipIfContainerRuntimeIs](https://github.com/kubernetes/kubernetes/blob/afc4506c33063fe64c840f2696af42bfb996436c/test/e2e/framework/util.go#L348) to make `SkipUnlessContainerRuntimeIs`.

This should work for rocket too, right?

File: test/e2e_node/security_context_test.go [hunk 9]
```diff
@@ -274,4 +276,84 @@ var _ = framework.KubeDescribe("Security Context", func() {
...
+
+			if err := podClient.MatchContainerOutput(podName, podName, output); err != nil {
+				return err
+			}
+
+			return nil
+		}
+
+		It("should allow privilege escalation when not explicitly set and uid != 0", func() {
+			podName := "alpine-nnp-" + string(uuid.NewUUID())
```
Comment (timstclair) [inline test/e2e_node/security_context_test.go]: nit: Just use the `GenerateName` on the pod spec.

File: pkg/securitycontext/util.go (hunk 7 above)
Comment (jessfraz) [inline pkg/securitycontext/util.go]: yeah, if the default is to use whatever the dockerfile is if not set thru the k8s communication with the docker API, then yes, I am not familiar with the behavior tho

File: pkg/security/podsecuritypolicy/provider.go (hunk 6 above)
Comment (jessfraz) [inline pkg/security/podsecuritypolicy/provider.go]: we don't have a default we set we just have a behavior though when it is nil (proposal: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/no-new-privs.md#pod-security-policy-changes), if I understand your comment correctly, so it is allowed to be nil here I think

File: test/images/nonewprivs/Makefile (hunk 8 above)
Comment (jessfraz) [inline test/images/nonewprivs/Makefile]: i think they are all google_containers? should they all change?

File: test/e2e_node/security_context_test.go (hunk 5 above)
Comment (jessfraz) [inline test/e2e_node/security_context_test.go]: this test works with rkt tho, wouldnt that make it skip it if it wasn't docker?

File: pkg/securitycontext/util.go [hunk 10]
```diff
@@ -232,3 +237,51 @@ func internalSecurityContextFromPodSecurityContext(pod *api.Pod) *api.SecurityCo
...
+	if sc.Capabilities != nil {
+		for _, cap := range sc.Capabilities.Add {
+			if string(cap) == "CAP_SYS_ADMIN" {
+				return false
+			}
+		}
+	}
+
+	// handle the case where defaultAllowPrivilegeEscalation is false or the user explicitly set allowPrivilegeEscalation to false
+	if sc.AllowPrivilegeEscalation != nil && !*sc.AllowPrivilegeEscalation {
```
Comment (timstclair) [inline pkg/securitycontext/util.go]: nit: simplify to:
```go
if sc.AllowPrivilegeEscalation != nil {
  return !*sc.AllowPrivilegeEscalation
}
```

File: pkg/security/podsecuritypolicy/provider.go (hunk 6 above)
Comment (timstclair) [inline pkg/security/podsecuritypolicy/provider.go]: Let me see if I can describe the scenario better:

1. Admin sets `psp.AllowPrivilegeEscalation = false`, but leaves `psp.DefaultAllowPrivilegeEscalation = nil`.
2. User creates a pod with a non-root user, and leaves `pod.Spec.SecurityContext.AllowPrivilegeEscalation = nil`.
3. Validation check here lets it through, because it is nil.
4. Kubelet defaults it to allowed

As an admin, I would expect that the value I set in (1) would enforce that `AllowPrivilegeEscalation` is always false (ignoring the privileged pod case), and the result in (4) would surprise me.

File: pkg/securitycontext/util.go (hunk 7 above)
Comment (timstclair) [inline pkg/securitycontext/util.go]: Yes, if the user isn't set through k8s, the value from the dockerfile is used. I know what you have here is consistent with the proposal, but I'm worried about breaking users with this. It seems like we're already turning it off in most of the cases where it would actually be useful as a security primitive, right? In that case, can we just leave it off in the default case? (i.e. if `pod.Spec..AllowPrivilegeEscalation == nil && psp.DefaultAllowPrivilegEscalation == nil` then the container is run without `no_new_privs`?)

File: test/images/nonewprivs/Makefile (hunk 8 above)
Comment (timstclair) [inline test/images/nonewprivs/Makefile]: Eh, you can leave it. We're gradually transitioning everything to `google-containers`, but we should probably just send a separate PR to fix all of these.

File: pkg/security/podsecuritypolicy/provider.go (hunk 6 above)
Comment (jessfraz) [inline pkg/security/podsecuritypolicy/provider.go]: ah I see

File: pkg/securitycontext/util.go (hunk 7 above)
Comment (jessfraz) [inline pkg/securitycontext/util.go]: yeah this makes sense

File: pkg/security/podsecuritypolicy/provider.go (hunk 6 above)
Comment (jessfraz) [inline pkg/security/podsecuritypolicy/provider.go]: so `psp.AllowPrivilegeEscalation` is just "am i allowed to set securityContext.AllowPrivilegeEscalation" so I think it might be a naming problem

Comment (jessfraz) [inline pkg/security/podsecuritypolicy/provider.go]: the naming convention was taken from psp.Privileged which I also agree is confusing

File: test/images/nonewprivs/Dockerfile (hunk 3 above)
Comment (jessfraz) [inline test/images/nonewprivs/Dockerfile]: this wont work if its not done like this

Comment (lucab) [inline test/images/nonewprivs/Dockerfile]: No worry, this was just a minor suggestion. Is the sticky bit dropped during dockerbuild COPY?

Comment (jessfraz) [inline test/images/nonewprivs/Dockerfile]: yeah i am assuming it's that didn't have time to debug

**Comment by jessfraz** [comment]
all green and rebased

File: pkg/kubelet/dockershim/security_context.go [hunk 11]
```diff
@@ -113,6 +113,14 @@ func modifyHostConfig(sc *runtimeapi.LinuxContainerSecurityContext, hostConfig *
 	}
 	hostConfig.SecurityOpt = append(hostConfig.SecurityOpt, apparmorSecurityOpts...)
 
+	if sc.NoNewPrivs {
+		// "no-new-privileges" was added in Docker 1.11
+		// which is API version 1.23
+		if apiVersion.GTE(semver.Version{Major: 1, Minor: 23}) {
```
Comment (tallclair) [inline pkg/kubelet/dockershim/security_context.go]: Should we block the pod from running if `no-new-privileges` can't be enforced? That's what we do for apparmor.  (https://github.com/kubernetes/kubernetes/blob/3fb2d8248d7f367758292ac0964cf0d33a8f4678/pkg/kubelet/kubelet.go#L812-L813)

File: staging/src/k8s.io/api/core/v1/types.go [hunk 12]
```diff
@@ -4613,6 +4613,15 @@ type SecurityContext struct {
 	// Default is false.
 	// +optional
 	ReadOnlyRootFilesystem *bool `json:"readOnlyRootFilesystem,omitempty" protobuf:"varint,6,opt,name=readOnlyRootFilesystem"`
+	// AllowPrivilegeEscalation controls whether a process can gain more
+	// privileges than it's parent process. This bool directly controls if
+	// the no_new_privs flag will be set on the container process.
+	// By default, AllowPrivilegeEscalation is false unless the container is
+	// 1) run as Privileged
+	// 2) has CAP_SYS_ADMIN
+	// 3) UID of the container is not 0
```
Comment (tallclair) [inline staging/src/k8s.io/api/core/v1/types.go]: I believe this is no longer the case?

**Review by tallclair — COMMENTED** [review]
Just 1 nit & 1 question, then LGTM.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 12 above)
Comment (jessfraz) [inline staging/src/k8s.io/api/core/v1/types.go]: wah good catch thought i got them alll

File: pkg/kubelet/dockershim/security_context.go (hunk 11 above)
Comment (jessfraz) [inline pkg/kubelet/dockershim/security_context.go]: oh interesting... that would be cool

Comment (jessfraz) [inline pkg/kubelet/dockershim/security_context.go]: As long as it doesn't break people I think it should be ok to do

Comment (jessfraz) [inline pkg/kubelet/dockershim/security_context.go]: So I am not sure exactly how we would do this other than checking the kernel version is >= 3.5

Comment (tallclair) [inline pkg/kubelet/dockershim/security_context.go]: It doesn't need to be perfect... You could start with `runtime == rkt || (runtime == docker && apiversion >= 1.23)`

**Comment by jessfraz** [comment]
It works in rkt :)

File: pkg/kubelet/dockershim/security_context.go (hunk 11 above)
Comment (jessfraz) [inline pkg/kubelet/dockershim/security_context.go]: updated

Comment (jessfraz) [inline pkg/kubelet/dockershim/security_context.go]: should I get rid of this check then?

File: pkg/kubelet/lifecycle/handlers.go [hunk 13]
```diff
@@ -165,3 +165,75 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
...
+		return PodAdmitResult{Admit: true}
+	}
+
+	// If the containers in a pod do not require no-new-privs, admit it.
+	if !noNewPrivsRequired(attrs.Pod) {
+		return PodAdmitResult{Admit: true}
+	}
+
+	// Make sure it is either docker or rkt runtimes.
+	if a.Runtime.Type() != kubetypes.DockerContainerRuntime && a.Runtime.Type() != kubetypes.RktContainerRuntime {
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: @yujuhong - It would be nice to incorporate this type of logic into the CRI. What if there was something like "GetSupportedFeatures", which would return a list of optional features (known constants) that the CRI runtime supported?

File: pkg/kubelet/lifecycle/handlers.go [hunk 14]
```diff
@@ -165,3 +165,75 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
...
+		return PodAdmitResult{
+			Admit:   false,
+			Reason:  "NoNewPrivs",
+			Message: fmt.Sprintf("Cannot enforce NoNewPrivs: %s runtime not supported", a.Runtime.Type()),
+		}
+	}
+
+	// Make sure docker api version is valid.
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: move this into the if block below, after you check whether the runtime is docker.

**Comment by spxtr** [comment]
Lint comments on generated code aren't too useful :disappointed:

**Comment by jessfraz** [comment]
this is all updated and green :)

**Comment by jessfraz** [comment]
ping @smarterclayton for approval 😇

**Comment by smarterclayton** [comment]
API approved as per the proposal and discussion on the thread.  

/approve

File: test/images/nonewprivs/nnp.c [hunk 15]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2017 The Kubernetes Authors.
```
Comment (mkumatag) [inline test/images/nonewprivs/nnp.c]: Is there any specif reason my we are using C code? I see even os library in go has `Geteuid()` which can be used.

Comment (jessfraz) [inline test/images/nonewprivs/nnp.c]: Does it matter?

Comment (mkumatag) [inline test/images/nonewprivs/nnp.c]: Yes. For multi arch test images we are planning to have all images under `kubernetes/tests/images` directory for supported architectures like `amd64, arm, arm64, ppc64le`. So its always very easy to cross build the go code over a C code.

**Comment by grodrigues3** [comment]
Which sig does this fall under?

**Comment by jessfraz** [comment]
sig-auth

File: pkg/apis/extensions/v1beta1/types.go (hunk 4 above)
Comment (php-coder) [reply pkg/apis/extensions/v1beta1/types.go]: @jessfraz I see that s/it's parent/its parent/ wasn't fixed. Do you want me to prepare PR for that?

**Comment by jessfraz** [comment]
Sure thanks

File: test/e2e_node/security_context_test.go (hunk 9 above)
Comment (php-coder) [reply test/e2e_node/security_context_test.go]: Unfortunately, it's impossible because we need to know a pod name to be able to wait for it and inspect its logs later.

File: pkg/security/podsecuritypolicy/provider_test.go [hunk 16]
```diff
@@ -1033,3 +1034,111 @@ func TestValidateAllowedVolumes(t *testing.T) {
...
 	}
 }
+
+// TestValidateAllowPrivilegeEscalation will test that when the podSecurityPolicy
+// AllowPrivilegeEscalation is false we cannot set a container's securityContext
+// to allowPrivilegeEscalation, but when it is true we can.
+func TestValidateAllowPrivilegeEscalation(t *testing.T) {
+	pod := defaultPod()
+	pe := true
+	pod.Spec.Containers[0].SecurityContext.AllowPrivilegeEscalation = &pe
```
Comment (php-coder) [inline pkg/security/podsecuritypolicy/provider_test.go]: I see that tests don't cover that case of using InitContainers? Does this validation work for them? I assume that it should but I don't see any mentions about them in the code.

File: pkg/kubelet/lifecycle/handlers.go [hunk 17]
```diff
@@ -165,3 +165,74 @@ func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
...
+				Reason:  "NoNewPrivs",
+				Message: fmt.Sprintf("Cannot enforce NoNewPrivs: docker runtime API version %q must be greater than or equal to 1.23", rversion.String()),
+			}
+		}
+	}
+
+	return PodAdmitResult{Admit: true}
+}
+
+func noNewPrivsRequired(pod *v1.Pod) bool {
```
Comment (php-coder) [inline pkg/kubelet/lifecycle/handlers.go]: @jessfraz Should we handle `InitContainers` here too? Looks like, yes, we should.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 18]
```diff
@@ -4613,6 +4613,14 @@ type SecurityContext struct {
 	// Default is false.
 	// +optional
 	ReadOnlyRootFilesystem *bool `json:"readOnlyRootFilesystem,omitempty" protobuf:"varint,6,opt,name=readOnlyRootFilesystem"`
+	// AllowPrivilegeEscalation controls whether a process can gain more
+	// privileges than it's parent process. This bool directly controls if
+	// the no_new_privs flag will be set on the container process.
+	// AllowPrivilegeEscalation is true always when the container is:
+	// 1) run as Privileged
+	// 2) has CAP_SYS_ADMIN
```
Comment (php-coder) [inline staging/src/k8s.io/api/core/v1/types.go]: @jessfraz Are we missing the case "when a container is not run as root" here?

Comment (jessfraz) [reply staging/src/k8s.io/api/core/v1/types.go]: We got rid of that functionality

File: pkg/kubelet/lifecycle/handlers.go (hunk 17 above)
Comment (jessfraz) [reply pkg/kubelet/lifecycle/handlers.go]: Are init containers handled differently that should probably be documented somewhere so it's not overlooked unless I missed it

File: staging/src/k8s.io/api/core/v1/types.go (hunk 18 above)
Comment (php-coder) [reply staging/src/k8s.io/api/core/v1/types.go]: And the proposal is outdated now? https://github.com/kubernetes/community/blob/f149433600ff1795521e7d9f2ae8ff3162394c61/contributors/design-proposals/no-new-privs.md#changes-of-securitycontext-objects

Comment (php-coder) [reply staging/src/k8s.io/api/core/v1/types.go]: Also some tests can be deleted then: https://github.com/kubernetes/kubernetes/blob/28f6b3fcc037c3a21048160d062f6d544081bd71/pkg/securitycontext/util_test.go#L203-L212

File: pkg/kubelet/lifecycle/handlers.go (hunk 17 above)
Comment (php-coder) [reply pkg/kubelet/lifecycle/handlers.go]: As far I understand, the most of the code from this PR is handling initContainers correctly because it works only with securityContext. The existing infrastructure is aware about containers/initContainers and executes validation on their SecurityContexts.

But when we're manually inspecting PodSpec we obligated to take care about initContainers. In this particular case, I see that `noNewPrivsRequired()` will return `false` for a pod that has initContainers with allowPrivielegeEscalation=false. I'm not sure what will be the consequence of this, perhaps `allowPrivielegeEscalation=false` won't work for a pod?

File: pkg/apis/extensions/v1beta1/types.go (hunk 4 above)
Comment (php-coder) [reply pkg/apis/extensions/v1beta1/types.go]: I've submitted https://github.com/kubernetes/kubernetes/pull/51536
